File: src/main/java/tn/cashfly/services/NewsService.java
package tn.cashfly.services;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import io.github.cdimascio.dotenv.Dotenv;

public class NewsService {

    private static final Dotenv dotenv = Dotenv.load();
    private static final String API_KEY = dotenv.get("NEWS_API_KEY");
    
    // Broadened query to ensure we get results: Tunisia OR Finance OR Economy in French
    private static final String NEWS_API_URL = "https://newsapi.org/v2/everything?q=%28finance+OR+%C3%A9conomie+OR+investissement%29+AND+%28tunisie+OR+maghreb%29&language=fr&sortBy=publishedAt&pageSize=15&apiKey="
            + API_KEY;

    public static class NewsArticle {
        public String title;
        public String description;
        public String url;
        public String publishedAt;
        public String source;

        public NewsArticle(String title, String description, String url, String publishedAt, String source) {
            this.title = title;
            this.description = description;
            this.url = url;
            this.publishedAt = publishedAt;
            this.source = source;
        }
    }

    public List<NewsArticle> fetchFinancialNews() {
        List<NewsArticle> articles = new ArrayList<>();
        try {
            if (API_KEY == null || API_KEY.isEmpty() || API_KEY.equals("your_newsapi_key_here")) {
                System.err.println("News API key is missing or invalid in .env");
                return getFallbackNews();
            }

            String response = makeGetRequest(NEWS_API_URL);
            if (response != null && !response.isEmpty()) {
                articles = parseNewsApiResponse(response);
            }
        } catch (Exception e) {
            System.err.println("News API fetch failed: " + e.getMessage());
        }

        // If API fails or returns no results, use high-quality fallback
        if (articles.isEmpty()) {
            System.out.println("Using fallback news articles...");
            articles = getFallbackNews();
        }
        return articles;
    }

    private String makeGetRequest(String apiUrl) throws Exception {
        URL url = new URL(apiUrl);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        conn.setConnectTimeout(8000);
        conn.setReadTimeout(8000);
        conn.setRequestProperty("User-Agent", "CashFly/1.0");
        conn.setRequestProperty("Accept", "application/json");
        int responseCode = conn.getResponseCode();
        if (responseCode == 200) {
            BufferedReader reader = new BufferedReader(
                    new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null)
                sb.append(line);
            reader.close();
            return sb.toString();
        }
        return null;
    }

    private List<NewsArticle> parseNewsApiResponse(String json) {
        List<NewsArticle> articles = new ArrayList<>();
        try {
            int articlesIndex = json.indexOf("\"articles\":");
            if (articlesIndex < 0) return articles;
            String[] items = json.substring(articlesIndex).split("\"source\":\\{");
            for (int i = 1; i < items.length && articles.size() < 10; i++) {
                String item = items[i];
                String title = extractJsonValue(item, "title");
                String description = extractJsonValue(item, "description");
                String url = extractJsonValue(item, "url");
                String publishedAt = extractJsonValue(item, "publishedAt");
                String source = extractJsonValue(item, "name");
                if (title != null && !title.isEmpty()) {
                    articles.add(new NewsArticle(title, description, url, publishedAt, source));
                }
            }
        } catch (Exception e) {
            System.err.println("Parse error: " + e.getMessage());
        }
        return articles;
    }

    private String extractJsonValue(String json, String key) {
        try {
            String searchKey = "\"" + key + "\":\"";
            int start = json.indexOf(searchKey);
            if (start < 0) return "";
            start += searchKey.length();
            int end = json.indexOf("\"", start);
            while (end > 0 && json.charAt(end - 1) == '\\') {
                end = json.indexOf("\"", end + 1);
            }
            if (end < 0) return "";
            return json.substring(start, end).replace("\\\"", "\"").replace("\\n", " ");
        } catch (Exception e) {
            return "";
        }
    }

    public List<NewsArticle> getFallbackNews() {
        List<NewsArticle> articles = new ArrayList<>();
        articles.add(new NewsArticle(
                "La Banque Centrale Tunisienne maintient son taux directeur à 8%",
                "Le Conseil d'Administration de la BCT a décidé de maintenir le taux d'intérêt directeur inchangé pour soutenir la stabilité financière.",
                "https://www.bct.gov.tn",
                "2024-03-01",
                "BCT Officiel"));
        articles.add(new NewsArticle(
                "Croissance du secteur des startups en Tunisie",
                "Le rapport annuel sur l'écosystème montre une augmentation de 15% des investissements en capital-risque cette année.",
                "https://www.smartcapital.tn",
                "2024-03-02",
                "Smart Capital"));
        articles.add(new NewsArticle(
                "Nouvelles mesures fiscales pour les PME",
                "Le ministère des Finances annonce des incitations pour les entreprises investissant dans le développement durable.",
                "https://www.finances.gov.tn",
                "2024-03-03",
                "Ministère des Finances"));
        return articles;
    }
}

File: src/Service/Api/NewsService.php
<?php

namespace App\Service\Api;

class NewsService
{
    private const API_KEY = '2a6e167a37374964b203cb3f15093dcc';
    private const API_URL = 'https://newsapi.org/v2/everything';

    public function getFinancialNews(int $limit = 10): array
    {
        $query = urlencode('(finance OR économie OR investissement OR Tunisia OR Maghreb)');
        $url = self::API_URL . "?q={$query}&language=fr&sortBy=publishedAt&pageSize={$limit}&apiKey=" . self::API_KEY;

        try {
            $response = $this->fetchUrl($url);
            $data = json_decode($response, true);

            if (isset($data['articles']) && is_array($data['articles'])) {
                return $this->formatArticles($data['articles']);
            }

            return $this->getFallbackNews();
        } catch (\Exception $e) {
            return $this->getFallbackNews();
        }
    }

    private function formatArticles(array $articles): array
    {
        $formatted = [];
        foreach ($articles as $article) {
            $formatted[] = [
                'title' => $article['title'] ?? 'Sans titre',
                'description' => $article['description'] ?? '',
                'url' => $article['url'] ?? '#',
                'source' => $article['source']['name'] ?? 'Unknown',
                'publishedAt' => isset($article['publishedAt']) ? $this->formatDate($article['publishedAt']) : '',
                'image' => $article['urlToImage'] ?? null,
            ];
        }
        return $formatted;
    }

    private function formatDate(string $date): string
    {
        $dateTime = new \DateTime($date);
        return $dateTime->format('d/m/Y H:i');
    }

    private function getFallbackNews(): array
    {
        return [
            [
                'title' => 'La BCT maintient son taux directeur inchangé',
                'description' => 'La Banque Centrale de Tunisie a décidé de maintenir son taux directeur à 8%, une décision qui vise à contenir l\'inflation tout en soutenant la croissance économique.',
                'url' => 'https://www.bct.gov.tn',
                'source' => 'Banque Centrale de Tunisie',
                'publishedAt' => date('d/m/Y'),
                'image' => null,
            ],
            [
                'title' => 'Smart Capital: Record d\'investissements étrangers en 2026',
                'description' => 'Smart Capital Tunisia annonce un record historique des investissements étrangers au premier trimestre 2026, avec une hausse de 25% par rapport à l\'année précédente.',
                'url' => 'https://www.smartcapital.tn',
                'source' => 'Smart Capital',
                'publishedAt' => date('d/m/Y'),
                'image' => null,
            ],
            [
                'title' => 'Ministère des Finances: Budget 2026 adopté',
                'description' => 'Le Parlement a adopté le projet de loi de finances 2026 qui prévoit une croissance de 4.5% et une inflation maîtrisée à 6%.',
                'url' => 'https://www.finances.gov.tn',
                'source' => 'Ministère des Finances',
                'publishedAt' => date('d/m/Y'),
                'image' => null,
            ],
            [
                'title' => 'Financement des PME: Nouvelles mesures annoncées',
                'description' => 'Le gouvernement annonce un package de 500 millions de dinars pour soutenir les PME tunisiennes dans leur développement.',
                'url' => 'https://www.tunisieindustrie.nat.tn',
                'source' => 'APII',
                'publishedAt' => date('d/m/Y'),
                'image' => null,
            ],
            [
                'title' => 'Bourse de Tunis: Performance record du Tunindex',
                'description' => 'L\'indice principal de la Bourse de Tunis a atteint un nouveau record historique, porté par les bonnes performances du secteur bancaire.',
                'url' => 'https://www.bvmt.com.tn',
                'source' => 'BVMT',
                'publishedAt' => date('d/m/Y'),
                'image' => null,
            ],
        ];
    }

    private function fetchUrl(string $url): string
    {
        $ch = curl_init();
        curl_setopt_array($ch, [
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => 30,
            CURLOPT_HTTPHEADER => [
                'Accept: application/json',
            ],
        ]);

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        if ($httpCode !== 200 || $response === false) {
            throw new \Exception('Failed to fetch news');
        }

        return $response;
    }
}
